doc/developer/design/20220516_transactional_consistency.md

```diff
@@ -0,0 +1,230 @@
+
+### Read Capabilities on Global Timestamp
+
+Proposal: Explicitly hold read capabilities for the global timestamp for all sources. All reads are assigned a timestamp
+equal to the global timestamp.
+
+Periodically the Coordinator will need to update the read capabilities to the current global timestamp. As the interval
+between read capability updates increase, the memory usage of each node goes up and the communication between each node
+goes down, and vice versa.
+
+This approach guarantees that when a read completes, the global timestamp is equal to or larger than the timestamp of
+that read.
+
+### Upstream Source Acknowledgements
+
+To achieve strict serializability of acknowledgements of upstream sources, acknowledgements should behave the same as a
+write transaction. The data being acknowledged is the write of the transaction and the acknowledgement is the commit.
+Materialize therefore must satisfy the following properties:
+
+1. If data with a timestamp `ts` has been acknowledged, then all reads must have a timestamp greater than or equal
+   to `ts`.
+2. If data with a timestamp `ts` has not been acknowledged, then no read can have a timestamp greater than or equal
+   to `ts`.
+
+Property 1 prevents the following scenario: a client goes to an upstream source and sees that some data has been
+acknowledged by Materialize but then can't see that data in Materialize.
+
+Property 2 prevents the following scenario: a client sees some data in Materialize, but doesn't see that it's been
+acknowledged in an upstream source.
+
+This guarantee is actually too strong for our needs and has very negative availability ramifications. Once an
+acknowledgement has been sent by Materialize, then no reads can be serviced until the upstream source has confirmed that
+it has received our acknowledgement. Before the confirmation, Materialize has no way of knowing if the acknowledgement
+has been received, and any timestamp selected for a read will be susceptible to one of the negative scenarios described
+above. This means is that if any upstream source is unresponsive to an acknowledgement, then all reads will be halted
+until that source is responsive again.
+
+Property 1 alone is still a useful property, because property 1 allows us to provide guarantees like the following: A
+client is using synchronous replication with an upstream source. When their transaction commits in the upstream source,
+they are guaranteed to also be able to see the transaction in Materialize. However, they may actually see the
+transaction in Materialize before their transaction commits in the upstream source. This is still a powerful and useful
+guarantee to provide. Therefore, we should remove the second bullet point from the correctness property and instead add
+the following correctness property:
+> Any write from an upstream source that has been acknowledged by Materialize will be visible in all future queries.
```

Comment:
is this only "if the user has opted into the stricter guarantees" as mentioned in the beginning of the document?

Also, is this enough? If we are not acknowledging writes in upstream, then for the real-time component of strict-serializability, we will also need to read what the latest write is from upstream, right?

Reply:
Ah yes, property #3. I need to think about why upstream acknowledgements make sense as part of property #3